src/data_storage/validators.py

```diff
@@ -0,0 +1,65 @@
+from typing import DefaultDict
+
+from common.entities.schemas import StoryItem
+from utilities.exceptions import (
+    ExistsCircleValidationError,
+    RootDoesNotExistValidationError,
+    UnconnectedNodeValidationError,
+    UnrelatedReferenceValidationError,
+)
+from utilities.functools import parse_graph_as_incidence_matrix
+
+
+def is_existing_graph_cycle(graph: DefaultDict[str, list], exit_nodes: set):
+    """
+    The basic idea is to start from node that has reference to None
+    or has no options(exit_node) and mark the node(that has exit and move)
+    to the adjacent unmarked node and continue this loop
+    until there is no unmarked adjacent node.
+
+    If we have unmarked node that means story has branch without end.
+    """
+
+    has_exit = set()
+
+    def deep_first_search(current_vertex):
+
+        has_exit.add(current_vertex)
+
+        for node in graph[current_vertex]:
+            if node not in has_exit:
+                deep_first_search(node)
```

Comment:
Generally recursion is faster, but not memory efficient. Could you redesign it with iteration (using stack)?